Add Charmed Kubeflow Operators to list of available Kubeflow distributions (#2578)
kubeflow/website: content/en/docs/distributions/charmed/install-kubeflow.md

+++

> title = "Installing Kubeflow"

Reviewer: ```suggestion
title = "Installing Kubeflow with Charmed Operators"
```

description = "Instructions for Kubeflow deployment with Kubeflow Charmed Operators"
weight = 10
+++

> This guide lists the steps necessary to install Kubeflow on any conformant Kubernetes, including AKS, EKS, GKE, Openshift and any kubeadm-deployed cluster,  provided that you have access to it via `kubectl`. 

Reviewer: Suggestion: 

- Maybe make it a bit more "active"/engaging wrt users
- Spell out GKE, EKS, AKS, GKE and provide links to their docs to ease UX friction
- "OpenShift"
- Add a link to kubeadm (for less experienced users)

e.g.

```suggestion
This guide will go through the steps you need to install Kubeflow on any conformant Kubernetes, including [Azure Kubernetes Service (AKS)](https://docs.microsoft.com/en-us/azure/aks/), [Amazon Elastic Kubernetes Service (EKS)](https://docs.aws.amazon.com/eks/index.html), [Google Kubernetes Engine (GKE)](https://cloud.google.com/kubernetes-engine/docs/), [OpenShift](https://docs.openshift.com), and any [kubeadm](https://kubernetes.io/docs/reference/setup-tools/kubeadm/)-deployed cluster (provided that you have access to it via `kubectl`). 
```

LMKWYT @RFMVasconcelos  👍

Author: This looks good!

Author: Thank you!!

#### 1. Install the Juju client

On Linux, install `juju` via [snap](https://snapcraft.io/docs/installing-snapd) with the following command:

```bash
snap install juju --classic
```

> Alternatively,  `brew install juju` on macOS or [download the Windows installer](https://launchpad.net/juju/2.8/2.8.5/+download/juju-setup-2.8.5-signed.exe). 

Reviewer: Suggestion:

```suggestion
If you have macOS, you can use [Homebrew](https://brew.sh) and type `brew install juju` in the command line. For Windows, download the Windows [installer for Juju](https://launchpad.net/juju/2.8/2.8.5/+download/juju-setup-2.8.5-signed.exe).
```

Author: Thank you for these improvements!

#### 2. Connect Juju to your Kubernetes cluster

> In order to operate workloads in your Kubernetes cluster with Juju, you have to add your cluster to the list of *clouds* in juju via the `add-k8s` command.

Reviewer: ```suggestion
To operate workloads in your Kubernetes cluster with Juju, you have to add the cluster to the list of *clouds* in Juju via the `add-k8s` command.
```

- Juju or juju -- to be consistent, let's use "Juju"
- "Your" repetition

> If your Kubernetes config file is in the standard location (`~/.kube/config` on Linux), and you only have one cluster, you can simply run:

Reviewer: ```suggestion
If your Kubernetes config file is in the default location (such as `~/.kube/config` on Linux) and you only have one cluster, you can simply run:
```

```bash
juju add-k8s myk8s

> ```

Reviewer: ```suggestion
```

```

If your kubectl config file contains multiple clusters, you can specify the appropriate one by name:

```bash
juju add-k8s myk8s --cluster-name=foo
```

> Finally, to use a different config file, you can set the `KUBECONFIG` environment variable to point to the relevant file. For example:

Reviewer: ```suggestion

Finally, to use a different config file, you can set the `KUBECONFIG` environment variable to point to the relevant file. For example:
```

```bash
KUBECONFIG=path/to/file juju add-k8s myk8s
```

> For more details, see the [Juju documentation](https://juju.is/docs/clouds).

Reviewer: ```suggestion

For more details, go to the [official Juju documentation](https://juju.is/docs/clouds).
```

#### 3. Create a controller

To operate workloads on your Kubernetes cluster, Juju uses controllers. You can create a controller with the  `bootstrap`  command:

```bash
juju bootstrap myk8s my-controller
```

This command will create a couple of pods under the `my-controller` namespace. You can see your controllers with the `juju controllers` command.

You can read more about controllers in the [Juju documentation](https://juju.is/docs/creating-a-controller).

#### 4. Create a model

A model in Juju is a blank canvas where your operators will be deployed, and it holds a 1:1 relationship with a Kubernetes namespace.

You can create a model and give it a name, e.g. `kubeflow`, with the `add-model` command, and you will also be creating a Kubernetes namespace of the same name:

```bash
juju add-model kubeflow
```

> You can list your models with the `juju models` command.

Reviewer: ```suggestion

You can list your models with the `juju models` command.
```

Author: I don't understand this change. Could you please clarify?

#### 5. Deploy Kubeflow

[note type="caution" status="MIN RESOURCES"]

> The minimum resources to deploy `kubeflow` are - 50Gb of disk, 14Gb of RAM and 2 CPUs - available to your Linux machine or VM. 

Reviewer: Suggestion:

```suggestion
To deploy `kubeflow`, you'll need at least 50Gb available of disk, 14Gb of RAM, and 2 CPUs available in your machine/VM.
```

(because macOS, Windows)

> If you have fewer resources, deploy `kubeflow-lite` or `kubeflow-edge`.

Reviewer: ```suggestion
If you don't have sufficient resources, you can instead deploy `kubeflow-lite` or `kubeflow-edge`.
```

Can you please add external links to `lite` and `edge` installation? They could be very useful IMO

Author: These links are outdated, we put down those docs and have only [this](https://charmed-kubeflow.io/docs/operators-and-bundles) one atm

Author: which is pointed to right after in the doc.

Author: I will keep that in mind and as we re-add that type of docs in charmed kubeflow I will happily point to it :)

[/note]

> Once you have a model, you can simply `juju deploy` any of the provided [Kubeflow bundles](https://charmed-kubeflow.io/docs/operators-and-bundles) into your cluster. For the Kubeflow lite bundle, run:

Reviewer: ```suggestion
Once you have a model, you can simply `juju deploy` any of the provided [Kubeflow bundles](https://charmed-kubeflow.io/docs/operators-and-bundles) into your cluster. For the _Kubeflow lite_ bundle, run:
```

(adding emphasis on _Kubeflow lite_ since "lite" is not capitalized)

Author: why not? :)

```bash
juju deploy kubeflow-lite
```

> **Congratulations, Kubeflow is now installing !**

Reviewer: Is this the official message output?

Author: Nope, it is a kind note ;)

> **Congratulations, Kubeflow is now installing !**

Reviewer: Suggestion:

`For the _Kubeflow lite_ bundle, run:...`

```suggestion
and your Kubeflow installation should begin!
```

(I'd advise to use **bold** text in **Notes** and **UI Elements**, as well as **Important stuff** but, of course, it's just a siggestion)

Author: This is ok! I'll accept your suggestion :) Thank you for all the recommendations!

You can observe your Kubeflow deployment getting spun-up with the command:

```bash
watch -c juju status --color
```

> #### 6. Add an RBAC role for istio

Reviewer: ```suggestion
#### 6. Add an RBAC role for Istio
```

https://istio.io

> Currently, in order to setup Kubeflow with Istio correctly, you need to provide the `istio-ingressgateway` operator access to Kubernetes resources. The following command will create the appropriate role:

Reviewer: ```suggestion
At the time of writing this guide, to set up Kubeflow with [Istio](https://istio.io) correctly, you need to provide the `istio-ingressgateway` operator access to Kubernetes resources. Use the  following command to create the appropriate role:
```

Author: Thank you for this improvement! :)

```bash
kubectl patch role -n kubeflow istio-ingressgateway-operator -p '{"apiVersion":"rbac.authorization.k8s.io/v1","kind":"Role","metadata":{"name":"istio-ingressgateway-operator"},"rules":[{"apiGroups":["*"],"resources":["*"],"verbs":["*"]}]}'
```

#### 7. Set URL in authentication methods 

> A final step to enable your Kubeflow dashboard access is to provide your dashboard public URL to dex-auth and oidc-gatekeeper via the following commands:

Reviewer: ```suggestion
Finally, you need to enable your Kubeflow dashboard access. Provide the dashboard's public URL to dex-auth and oidc-gatekeeper as follows:
```

(I hope I got this right)

```bash
juju config dex-auth public-url=http://<URL>
juju config oidc-gatekeeper public-url=http://<URL>
```

> Where `<URL>` is the hostname that the Kubeflow dashboard responds to. 

Reviewer: ```suggestion
where in place of `<URL>` you should use the hostname that the Kubeflow dashboard responds to.
```

🤷

#### More documentation

> For more documentation visit [Charmed Kubeflow docs](https://charmed-kubeflow.io/docs)

Reviewer: ```suggestion
For more information, visit the [Charmed Kubeflow docs](https://charmed-kubeflow.io/docs).
```

Reviewer: How about this:

```suggestion
For more documentation, visit the [Charmed Kubeflow](https://charmed-kubeflow.io/docs).
```

("docs...docs" repetition)

Reviewer: Good idea [UPDATED above] 👍 😃

Author: Thank you!!

#### Having issues?

> If you face any difficulties following these instructions, please create an issue [here](https://github.com/juju-solutions/bundle-kubeflow/issues).

Reviewer: ```suggestion

If you have any issues or questions, feel free to create a GitHub issue [here](https://github.com/juju-solutions/bundle-kubeflow/issues).
```
